cluster-operator/src/main/java/io/strimzi/operator/cluster/operator/assembly/ConnectBuildOperator.java

```diff
@@ -0,0 +1,361 @@
+import io.strimzi.operator.common.Util;
+import io.strimzi.operator.common.operator.resource.BuildConfigOperator;
+import io.strimzi.operator.common.operator.resource.BuildOperator;
+import io.strimzi.operator.common.operator.resource.ConfigMapOperator;
+import io.strimzi.operator.common.operator.resource.DeploymentOperator;
+import io.strimzi.operator.common.operator.resource.PodOperator;
+import io.strimzi.operator.common.operator.resource.ServiceAccountOperator;
+import io.vertx.core.Future;
+
+import java.util.List;
+import java.util.concurrent.atomic.AtomicReference;
+
+public class ConnectBuildOperator {
+
+    private static final ReconciliationLogger LOGGER = ReconciliationLogger.create(ConnectBuildOperator.class.getName());
+
+    private final DeploymentOperator deploymentOperations;
+    private final PodOperator podOperator;
+    private final ConfigMapOperator configMapOperations;
+    private final ServiceAccountOperator serviceAccountOperations;
+    private final BuildConfigOperator buildConfigOperator;
+    private final BuildOperator buildOperator;
+
+    private final ImagePullPolicy imagePullPolicy;
+    private final List<LocalObjectReference> imagePullSecrets;
+    private final long connectBuildTimeoutMs;
+    private final PlatformFeaturesAvailability pfa;
+
+    private Reconciliation reconciliation;
+    private String namespace;
+    private KafkaConnectBuild connectBuild;
+
+    public ConnectBuildOperator(PlatformFeaturesAvailability pfa, ResourceOperatorSupplier supplier, ClusterOperatorConfig config) {
+        this.deploymentOperations = supplier.deploymentOperations;
+        this.podOperator = supplier.podOperations;
+        this.configMapOperations = supplier.configMapOperations;
+        this.serviceAccountOperations = supplier.serviceAccountOperations;
+        this.buildConfigOperator = supplier.buildConfigOperations;
+        this.buildOperator = supplier.buildOperations;
+
+        this.imagePullPolicy = config.getImagePullPolicy();
+        this.imagePullSecrets = config.getImagePullSecrets();
+        this.connectBuildTimeoutMs = config.getConnectBuildTimeoutMs();
+        this.pfa = pfa;
+    }
+
+    /**
+     * Builds a new container image with connectors on Kubernetes using Kaniko or on OpenShift using BuildConfig
+     *
+     * @param reconciliation    The reconciliation
+     * @param namespace         Namespace of the Connect cluster
+     * @param connectName       Name of the Connect cluster
+     * @param connectBuild      KafkaConnectBuild object
+     * @return                  Future for tracking the asynchronous result of the Kubernetes image build
+     */
+    public Future<BuildInfo> build(Reconciliation reconciliation, String namespace, String connectName, KafkaConnectBuild connectBuild) {
+        this.reconciliation = reconciliation;
+        this.namespace = namespace;
+        this.connectBuild = connectBuild;
+        return deploymentOperations.getAsync(namespace, connectName)
```

Comment:
If I read this code correctly, you seem to always get the Deployment. But if `connectBuild.getBuild() == null` you actually throw it all away. Would it make more sense to re-order the things and first check if the `onnectBuild.getBuild() != null` and only then pull the Deployment? It might safe an unnecessary call to the Kube APi in many cases where the build is not used.

Reply:
@scholzj  Yeah, you're right, I've done an additional update to move the deployment get.